### DISCUSSION TITLE
"juicefs mount -d ..." hanged when no permission to exec fusermount

### DISCUSSION THREAD

**Discussion by myc724** [juicedata/juicefs#264] [Q&A] [answered]
I’m building a flow which deploys juicefs as regular user (ie. no root access). In my flow, each remote process mounts volume as follow (from various machines). 

    juicefs mount -d redis://<meta_data_server>:<port_number>/1 /tmp/runlog

However, for some machine where the access permission of fusermount is not set properly (as follow for example), “juicefs mount -d …” hanged forever.

    -rwsr-x---. 1 root fuse 27968 Dec  7  2011 /user/bin/fusermount

In fact, juicefs without “-d” properly returned 1 and with the following error message.

    2021/03/17 23:41:20.331044 juicefs[31408] <FATAL>: fuse: fuse: fork/exec /usr/bin/fusermount: permission denied

I wonder whether the hanging of “juicefs mount -d …” is expected. If so, I will build my flow without “-d” then put the mounting process in the background ourselves. Let me know. Thanks!

**Comment by chnliyong** [comment, marked as answer]
It seems you OS is a a little bit old and the `fusermount` should be run by the `fuse` group. 
Non-privileged user mount will execute the mount by `fusermount` as it has set-user-ID bit, thus get the privilege.

Currently we check if the mount point is ready in a infinite loop,  we'll consider resolving this.

If you flow is executed with a specific regular user, you can add it to group `fuse` to bypass.

**Reply by myc724** [reply]
I see. Thanks! One quick follow-up question. The flow that I’m building could be used on public cloud. Is it proper-to/commonly expect any machine (on public cloud) has had fusermount installed and set as one of the following? Thanks!

Allow any user to execute:

    -rwsr-xr-x 1 root root 32096 Oct 30  2018 /usr/bin/fusermount

Only for fuse:

    -rwsr-x---. 1 root fuse 27968 Dec  7  2011 /usr/bin/fusermount

**Reply by chnliyong** [reply]
I've checked _Ubuntu_ `14.04, 16.04, 18.04, 20.04`, the `fusermount`'s mode is the same as the first above.
Also checked the fuse rpm package from [CentOS 7](https://centos.pkgs.org/7/centos-x86_64/fuse-2.9.2-11.el7.x86_64.rpm.html) and [CentOS 8](https://centos.pkgs.org/8/centos-baseos-x86_64/fuse-2.9.7-12.el8.x86_64.rpm.html), it's the same as the first one above.
```
$ rpm -qvlp fuse-2.9.2-11.el7.x86_64.rpm
warning: fuse-2.9.2-11.el7.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID f4a80eb5: NOKEY
-rw-r--r--    1 root     root                       38 Oct 30  2018 /etc/fuse.conf
-rwsr-xr-x    1 root     root                    32096 Oct 31  2018 /usr/bin/fusermount
-rwxr-xr-x    1 root     root                    15472 Oct 31  2018 /usr/bin/ulockmgr_server
-rwxr-xr-x    1 root     root                    11368 Oct 31  2018 /usr/sbin/mount.fuse
...

$ rpm -qvlp fuse-2.9.7-12.el8.x86_64.rpm
warning: fuse-2.9.7-12.el8.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID 8483c65d: NOKEY
-rwsr-xr-x    1 root     root                    38680 May 11  2019 /usr/bin/fusermount
-rwxr-xr-x    1 root     root                    17544 May 11  2019 /usr/bin/ulockmgr_server
drwxr-xr-x    2 root     root                        0 May 11  2019 /usr/lib/.build-id
```

It seems early OS's `fusermount` limit access only to `fuse` group, recently have no limit (**Not absolutely sure**).

**Reply by myc724** [reply]
Thanks for this information!

**Comment by davies** [comment]
@myc724 We have added a timeout for mounting in background, it will fail (exit with 1) if the mount point is not ready in 10 seconds.

**Reply by myc724** [reply]
Great! Thanks!
